pkg/controllers/cluster/cluster_controller.go

```diff
@@ -194,3 +244,153 @@ func (c *Controller) createExecutionSpace(cluster *v1alpha1.Cluster) error {
 
 	return nil
 }
+
+func (c *Controller) monitorClusterHealth() error {
+	clusterList := &v1alpha1.ClusterList{}
+	if err := c.Client.List(context.TODO(), clusterList); err != nil {
+		return err
+	}
+
+	clusters := clusterList.Items
+	for i := range clusters {
+		cluster := &clusters[i]
+		if err := wait.PollImmediate(MonitorRetrySleepTime, MonitorRetrySleepTime*HealthUpdateRetry, func() (bool, error) {
+			// Cluster object may be changed in this function.
+			if err := c.tryUpdateClusterHealth(cluster); err == nil {
+				return true, nil
+			}
+			clusterName := cluster.Name
+			if err := c.Get(context.TODO(), client.ObjectKey{Name: clusterName}, cluster); err != nil {
+				// If the cluster does not exist any more, we delete the health data from the map.
+				if apierrors.IsNotFound(err) {
+					c.clusterHealthMap.Delete(clusterName)
+					return true, nil
+				}
+				klog.Errorf("Getting a cluster to retry updating cluster health error: %v", clusterName, err)
+				return false, err
+			}
+			return false, nil
+		}); err != nil {
+			klog.Errorf("Update health of Cluster '%v' from Controller error: %v. Skipping.", cluster.Name, err)
+			continue
+		}
+	}
+
+	return nil
+}
+
+// tryUpdateClusterHealth checks a given cluster's conditions and tries to update it.
```

Comment:
```suggestion
// tryUpdateClusterHealth checks a given cluster's conditions and tries to update it.
//nolint:gocyclo
```
Let's ignore the `gocyclo` check for the moment. And improve it later.